fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/query/plan/plans/FDBComparatorPlanTest.java

```diff
@@ -0,0 +1,314 @@
+                .build();
+        // Note the reference index is now 1
+        RecordQueryPlan planUnderTest = RecordQueryComparatorPlan.from(plan(query1, query2), primaryKey(), 1);
+
+        int count = querySimpleRecordStore(NO_HOOK, planUnderTest, EvaluationContext::empty,
+                record -> assertThat(record.getNumValue2(), is(2)),
+                context -> assertDiscardedAtMost(134, context));
+        assertEquals(33, count);
+    }
+
+    @DualPlannerTest
+    void testTwoDifferentInnerPlansOneEndsSooner() throws Exception {
+        complexQuerySetup(NO_HOOK);
+
+        RecordQuery query1 = RecordQuery.newBuilder()
+                .setRecordType("MySimpleRecord")
+                .setFilter(Query.field("num_value_2").equalsValue(1))
+                .build();
+        // This plan has same records as the previous one, but end sooner
+        RecordQuery query2 = RecordQuery.newBuilder()
+                .setRecordType("MySimpleRecord")
+                .setFilter(Query.and(
+                        Query.field("num_value_2").equalsValue(1),
+                        Query.field("rec_no").lessThan(5L)))
+                .build();
+
+        RecordQueryPlan planUnderTest = RecordQueryComparatorPlan.from(plan(query1, query2), primaryKey(), 0);
+
+        int count = querySimpleRecordStore(NO_HOOK, planUnderTest, EvaluationContext::empty,
+                record -> assertThat(record.getNumValue2(), is(1)),
+                context -> assertDiscardedAtMost(134, context));
+        assertEquals(33, count);
+    }
+
+    @DualPlannerTest
+    void testTwoDifferentInnerPlansReferenceEndsSooner() throws Exception {
+        complexQuerySetup(NO_HOOK);
+
+        RecordQuery query1 = RecordQuery.newBuilder()
+                .setRecordType("MySimpleRecord")
+                .setFilter(Query.field("num_value_2").equalsValue(1))
+                .build();
+        // This plan has same records as the previous one, but end sooner
+        RecordQuery query2 = RecordQuery.newBuilder()
+                .setRecordType("MySimpleRecord")
+                .setFilter(Query.and(
+                        Query.field("num_value_2").equalsValue(1),
+                        Query.field("rec_no").lessThan(5L)))
+                .build();
+
+        // Note this time the reference plan is hte second one
+        RecordQueryPlan planUnderTest = RecordQueryComparatorPlan.from(plan(query1, query2), primaryKey(), 1);
+
+        int count = querySimpleRecordStore(NO_HOOK, planUnderTest, EvaluationContext::empty,
+                record -> assertThat(record.getNumValue2(), is(1)),
+                context -> assertDiscardedAtMost(10, context));
+        assertEquals(2, count);
+    }
+
+    @DualPlannerTest
+    void testNoInnerPlansFail() throws Exception {
+        complexQuerySetup(NO_HOOK);
+
+        assertThrows(VerifyException.class, () -> RecordQueryComparatorPlan.from(Collections.emptyList(), primaryKey(), 0));
```

Comment:
minor: there's part of me that wonders if this should be a `RecordCoreArgumentException`, partially to make sure that as many errors as possible are in the `RecordCoreException` hierarchy, but I suppose it's a minor thing

Reply:
I was following patterns I've seen around (i.e. using `Verify.verify` for parameter verification. I don't mind changing them to `RecordCoreArgumentException` if that's the way to go.

Reply:
Done.